clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -343,15 +343,34 @@ static void collectSYCLAttributes(Sema &S, FunctionDecl *FD,
   if (!FD->hasAttrs())
     return;
 
-  llvm::copy_if(FD->getAttrs(), std::back_inserter(Attrs), [](Attr *A) {
-    // FIXME: Make this list self-adapt as new SYCL attributes are added.
-    return isa<IntelReqdSubGroupSizeAttr, IntelNamedSubGroupSizeAttr,
-               ReqdWorkGroupSizeAttr, SYCLIntelKernelArgsRestrictAttr,
-               SYCLIntelNumSimdWorkItemsAttr,
-               SYCLIntelSchedulerTargetFmaxMhzAttr,
-               SYCLIntelMaxWorkGroupSizeAttr, SYCLIntelMaxGlobalWorkDimAttr,
-               SYCLIntelNoGlobalWorkOffsetAttr, SYCLSimdAttr>(A);
-  });
+  // Attributes that should be propagated from device functions to a kernel
+  // in SYCL 1.2.1.
+  if (S.getASTContext().getLangOpts().getSYCLVersion() <
+      LangOptions::SYCL_2020) {
+    llvm::copy_if(FD->getAttrs(), std::back_inserter(Attrs), [](Attr *A) {
+      // FIXME: Make this list self-adapt as new SYCL attributes are added.
+      return isa<SYCLIntelKernelArgsRestrictAttr, SYCLIntelNumSimdWorkItemsAttr,
```

Comment:
`IntelNamedSubGroupSizeAttr` is missing here

Reply:
> `IntelNamedSubGroupSizeAttr` is missing here

This is sycl2020 attribute, this should not be propagated to earlier sycl mode. @erichkeane, could you please confirm this?

Reply:
Named sub group size is supposed to be 2020 only.  Note that ReqdSubGroupSize is 'both' based on its spelling, so there might be some additional work that needs to happen.

Reply:
Do we have a diagnostic when this attribute is used in earlier versions?

Reply:
IntelNamedSubGroupSizeAttr is sycl2020 attribute. i see that all Lit tests (sub-group-size.cpp) is passing the -sycl-std=2020 flag and we are not using this in earlier sycl version

Reply:
> IntelNamedSubGroupSizeAttr is sycl2020 attribute.

I don't see it listed in the SYCL 2020 spec?

Reply:
Inside IntelNamedSubGroupSizeDocs, i see that this attribute follows SYCL 2020 Attribute Rules. 

// Check SYCL2020 Attributes.  2020 attributes don't propagate, 
static void CheckSYCL2020Attributes(
    Sema &S, FunctionDecl *SYCLKernel, FunctionDecl *KernelBody,
    const llvm::SmallPtrSetImpl<FunctionDecl *> &CalledFuncs) {

  if (KernelBody) {
    // Make sure the kernel itself has all the 2020 attributes, since we don't
    // do propagation of these.
    if (auto *A = KernelBody->getAttr<IntelReqdSubGroupSizeAttr>())
      if (A->isSYCL2020Spelling())
        SYCLKernel->addAttr(A);
    if (auto *A = KernelBody->getAttr<IntelNamedSubGroupSizeAttr>())
      SYCLKernel->addAttr(A);

Reply:
My understanding (and maybe I'm wrong) is that attributes named `[[cl::whatever]]` or `__attribute__((whatever))` are SYCL 2017 (1.2.1) attributes, ones named `[[sycl::whatever]]` are SYCL 2020 attributes, and ones named `[[intel::whatever]]` are Intel-specific attributes. I think the SYCL 2017 and SYCL 2020 attributes need to do whatever the SYCL spec says, but the Intel-specific attributes may have custom behavior.

If that's correct, then I think the Intel-specific attributes should be handled on a case-by-case basis but I'd assume the default policy is that they behave however we want them to behave (and document that behavior) and that behavior is fixed & independent of whatever SYCL spec is currently being targeted. So it should continue to do whatever it used to do because users are likely relying on the documented behavior and we don't want to silently change the meaning of their code (unless we're fixing an actual bug).

Reply:
From a quick look at the implementation of `IntelNamedSubGroupSizeAttr`, I THINK prior to this patch, `IntelNamedSubGroupSizeAttr` was 'collected' in `collectSYCLAttributes ` but not added to kernel in `PropagateAndDiagnoseDeviceAttr`.  I do not know why the attribute was 'collected' in the first place because diagnostics for this attribute is handled in `CheckSYCL2020Attributes`, which as far as I can tell doesn't use collected attribute set. I suspect with this new patch (where we no longer 'collect' attribute) we can remove corresponding case in `PropagateAndDiagnoseDeviceAttr`

Also, I may be misunderstanding something but I THINK `IntelNamedSubGroupSizeAttr` is added irrespective of sycl version. This might be an existing bug ? @erichkeane

Reply:
Do you mean this seciton of code?  Those were only 'collected' when they were 'DirectlyCalled', which means we 'collect' the ones that were 'on the functor', not all children.  

We DO skip handling it otherwisein PropagateAndDiagnoseDeviceAttr, since it is supposed to be the SYCL1.2.1 attribute handling locations, an d CheckSYCL2020Attributes does the 2020 ones.

Reply:
Yes but is there any specific reason for why we were 'collecting' SYCL2020 attributes in this section of code? I see 'collected' attributes being handled only in `PropagateAndDiagnoseDeviceAttr` which like you mentioned, doesn't handle SYCL2020 attributes. Am I missing something?

Reply:
I think they still need to be 'collected' in order to work in that loop that does the checks.  We are in an interesting case where the same attribute (depending on spelling) can be either a 1.2.1 spelling or a 2020 spelling, so I think we collected them there.  The idea is that 'collected' ones that aren't needed are ignored in `PropagateAndDiagnoseDeviceAttr` (see the `isSYCL2020Spelling` for `ReqdWorkGroupSizeAttr` there), and their 2020 spellings are handled in the `CheckSYCL2020Attributes` function, which also uses the collecting.

Reply:
I was talking about `IntelNamedSubGroupSizeAttr`, not `ReqdWorkGroupSizeAttr` :) I was trying to understand if removing `IntelNamedSubGroupSizeAttr` from this section of code (as this new patch has done for pre-SYCL2020) has any implications I'm unaware of. 

Also, probably should research this first lol - I don't really know what SYCL2020 attributes mean. I haven't verified this but from the code it looks like `IntelNamedSubGroupSizeAttr ` is handled irrespective of SYCL version. Should this attribute even be available in earlier versions of SYCL?

Reply:
Given that it's an extension and not part of SYCL, I think we get to decide what versions of SYCL we want to support it in. Does the attribute have sensible semantics we can apply in SYCL 1.2.1? If so, then I'd say it should follow the usual SYCL 1.2.1 propagation behavior in that mode.

Reply:
>>Does the attribute have sensible semantics we can apply in SYCL 1.2.1? 

@AaronBallman, The attribute does not seem follow any semantics that can apply in SYCL 1.2.1

Please see here: https://github.com/intel/llvm/blob/sycl/clang/include/clang/Basic/AttrDocs.td#L4487

Reply:
If it doesn't have sensible semantics in 1.2.1, then I'd say we should diagnose using it in that mode as an ignored attribute (add a new `LangOpt` for SYCL >= 2020 mode, then use it in the attribute definition).